### DISCUSSION TITLE
How to resize the existing OSD

### DISCUSSION THREAD

**Discussion by prabhinmp** [rook/rook#10214] [General]
HI,
I have 3 nodes where 50Gb raw disk is attached and used by ceph , configured with replicapool of size 3.
So all the data on the node1 disk is replicated to node2 and node3.

Now the situation is 50Gb is getting filled up and need to resize  50Gb to 100GB. Will rook ceph support this?

**Comment by travisn** [comment]
Yes, if you increase the disk size, the OSD will notice the change the next time the pod restarts and Ceph should then report the new size available.

**Reply by ghost** [reply]
Once expanded the OSD, don't forget to tell bluestore about this, otherwise the newly added space will get reported as `RAW SIZE` only, without actually letting you use it as available/free space.

```
ceph-bluestore-tool bluefs-bdev-expand --path /var/lib/ceph/osd/ceph-<ID>
```

> Make sure to run this only when OSD is down! (Can be achieved by `kubectl -n rook-ceph scale deployment rook-ceph-osd-<ID> --replicas=0`); Don't forget to set it back to `1` replicas).  
> To run `ceph-bluestore-tool bluefs-bdev-expand ...` I had to `kubectl -n rook-ceph get deployment rook-ceph-osd-<ID> -o yaml > resize-osd-<ID>.yaml` and modify it significantly (into the `Job` kind, to make sure it runs only that command). I hope the official rook-ceph will support this functionality.

Ignore these errors:

```
2022-08-28T21:00:43.313+0000 7f59c45ad700 -1 bluestore(/var/lib/ceph/osd/ceph-<ID>) _read_bdev_label failed to read from /var/lib/ceph/osd/ceph-<ID>: (21) Is a directory
2022-08-28T21:00:43.313+0000 7f59c45ad700 -1 bluestore(/var/lib/ceph/osd/ceph-<ID>) unable to read label for /var/lib/ceph/osd/ceph-<ID>: (21) Is a directory
```

The reason for these errors is explained [here](https://youtu.be/e0J0gft2jEY?t=779)

The same in [this](https://github.com/rook/rook/issues/2997#issuecomment-1229554142) thread.

**Reply by travisn** [reply]
Rook adds an init container to the OSD deployments to run the `ceph-bluestore-tool bluefs-bdev-expand` every time the OSD starts, to ensure the resize will be picked up.

**Reply by ghost** [reply]
> Rook adds an init container to the OSD deployments to run the `ceph-bluestore-tool bluefs-bdev-expand` every time the OSD starts, to ensure the resize will be picked up.

yeah, I've noticed it is there but is only for `raw` devices, not for the `lvm`.
https://github.com/rook/rook/blob/a730d6d8019816590207c108bf935af0c034885d/pkg/operator/ceph/cluster/osd/spec.go#L501-L528

In my case I've extended the backing LV using the `lvextend` command.

Do you now if I can make it somehow to assume the backing device is `raw` for a short period of time, so it runs the `getExpandPVCInitContainer` bit which does the `ceph-bluestore-tool bluefs-bdev-expand`, and then switching it back to the `lvm`? Or that would not be safe?

**Reply by travisn** [reply]
There are a couple of options:
1. Add the init container manually to the deployment spec and let it resize when it restarts the OSD. The next time Rook reconciles the cluster, it will again remove the init container automatically.
2. Use the [Krew plugin "debug" command](https://github.com/rook/kubectl-rook-ceph#debug-mode) to initialize the OSD without running the daemon, then you could connect to the pod and run the expand command manually.

**Reply by travisn** [reply]
And actually, we should support the osd resize for lvm-based OSDs, especially if we just need to add the resize init container. @satoru-takeuchi Any concerns with that?

**Reply by travisn** [reply]
@andy108369 Want to open an issue for this, instead of this discussion?

**Reply by ghost** [reply]
> @andy108369 Want to open an issue for this, instead of this discussion?

Will do

**Reply by ghost** [reply]
=> https://github.com/rook/rook/issues/10835

**Reply by satoru-takeuchi** [reply]
@travisn No, there isn't any concern. This way should work.

**Reply by samos667** [reply]
> > Rook adds an init container to the OSD deployments to run the `ceph-bluestore-tool bluefs-bdev-expand` every time the OSD starts, to ensure the resize will be picked up.
> 
> yeah, I've noticed it is there but is only for `raw` devices, not for the `lvm`.
> 
> https://github.com/rook/rook/blob/a730d6d8019816590207c108bf935af0c034885d/pkg/operator/ceph/cluster/osd/spec.go#L501-L528
> 
> In my case I've extended the backing LV using the `lvextend` command.
> 
> Do you now if I can make it somehow to assume the backing device is `raw` for a short period of time, so it runs the `getExpandPVCInitContainer` bit which does the `ceph-bluestore-tool bluefs-bdev-expand`, and then switching it back to the `lvm`? Or that would not be safe?

How the type (LVM or RAW) is determined ? My nodes k8s are virtualized with proxmox and have a dedicated RAW virtual disk. Nodes access to the disk via /dev/vdb. And rook use this for make a OSD. But when i expand the disk in proxmox for 50GB, after restarting OSD the pool of data is incremented by 50GB but their are considered used. And if i have uderstood correctly i need to use blue-fs to extend the OSD correcty ? But having RAW disks, I don't understand why it's not done automatically when restarting the OSD.

**Comment by satoru-takeuchi** [comment]
@samos667 It looks another problem. Could you open a new issue?

**Reply by samos667** [reply]
It's done: https://github.com/rook/rook/issues/10930
I thought my problem was the same as this: https://github.com/rook/rook/issues/2997#issuecomment-984213483
@andy108369 replied to this comment, stating that you have to use bluestore to resize the OSD, however when I add an initcontainer to my deployment as mentionned here: https://github.com/rook/rook/issues/2997#issuecomment-1237966127 , the OSD becomes completely broken
